examples/reporting/parallel_report_download.rb

```diff
@@ -0,0 +1,186 @@
+
+require 'optparse'
+require 'google/ads/google_ads'
+require 'thread'
+
+def parallel_report_download(customer_ids, login_customer_id)
+  # GoogleAdsClient will read a config file from
+  # ENV['HOME']/google_ads_config.rb when called without parameters
+  client = Google::Ads::GoogleAds::GoogleAdsClient.new
+
+  # Optional login customer ID if your access to the CIDs is
+  # via a manager account.
+  client.configure do |config|
+    if login_customer_id
+      config.login_customer_id = login_customer_id.tr("-", "").to_i
+    end
+  end
+
+  ga_service = client.service.google_ads
+
+  query_list = {}
+  query_list["Campaign Query"] = <<~QUERY
+    SELECT campaign.id, metrics.impressions, metrics.clicks
+    FROM campaign
+    WHERE segments.date DURING LAST_30_DAYS
+  QUERY
+  query_list["Ad Group Query"] = <<~QUERY
+    SELECT campaign.id, ad_group.id, metrics.impressions, metrics.clicks
+    FROM ad_group
+    WHERE segments.date DURING LAST_30_DAYS
+  QUERY
+
+  reports_succeeded = Queue.new()
+  reports_failed = Queue.new()
+
+  # Start all the threads.
+  # This is a naive implementation for illustrative purposes. It is possible to
+  # optimize the utilization of each customer ID by providing a queue of work
+  # (or similar). However, this would complicate the example code and so is
+  # omitted here.
+  threads = []
+  query_list.each do |query_key, query|
+    customer_ids.each do |cid|
+      cid = cid.tr("-", "")
+      # Starts the report download in a background thread.
+      threads << Thread.new do
+        begin
+          puts "Requesting #{query_key} for CID #{cid}"
+          responses = ga_service.search_stream(
+            customer_id: cid.tr("-", ""),
+            query: query,
+          )
+          # Stores the number of rows for each report for illustrative purposes.
+          # Users of this code example can implement other logic here such as
+          # storing response to a database/file, pass the response on to
+          # another method for further processing, etc.
+          num_rows = 0
+          responses.each do |response|
+            response.results.each do |row|
+              num_rows += 1
+            end
+          end
+          reports_succeeded << {
+            :cid => cid,
+            :query_key => query_key,
+            :num_rows => num_rows,
+          }
+        rescue Google::Ads::GoogleAds::Errors::GoogleAdsError => e
+          error_messages = ""
+          e.failure.errors.each do |error|
+            error_messages += error.message
+          end
+          reports_failed << {
+            :cid => cid,
+            :query_key => query_key,
+            :error_messages => error_messages,
+          }
+        end
+      end
+    end
+  end
+
+  puts "Awaiting results from #{threads.size} report download requests..."
+
+  # Waits for all pending requests to the current set of customer IDs
+  # to complete.
+  threads.each { |aThread|  aThread.join }
```

Comment:
Just call this "thread" instead of "aThread"

Reply:
Fixed.